synapse/handlers/oidc_handler.py

```diff
@@ -0,0 +1,452 @@
+        args = {
+            "grant_type": "authorization_code",
+            "code": code,
+            "redirect_uri": self._callback_url,
+        }
+
+        uri, headers, body = self._client_auth.prepare(
+            method="POST", uri=token_endpoint, headers={}, body=""
+        )
+        headers = {k: [v] for (k, v) in headers.items()}
+        qs = parse_qsl(body, keep_blank_values=True)
+        args.update(qs)
+
+        try:
+            resp = await self._http_client.post_urlencoded_get_json(
+                uri, args, headers=headers
+            )
+        except HttpResponseException as e:
+            resp = json.loads(e.response)
+
+        if "error" not in resp:
+            return resp
+
+        error = resp["error"]
+        description = resp.get("error_description", error)
+        raise OidcError(error, description)
+
+    async def _fetch_userinfo(self, token: Dict[str, str]) -> UserInfo:
+        metadata = await self.load_metadata()
+
+        resp = await self._http_client.get_json(
+            metadata["userinfo_endpoint"],
+            headers={"Authorization": ["Bearer {}".format(token["access_token"])]},
+        )
+
+        return UserInfo(resp)
+
+    async def _parse_id_token(self, token: Dict[str, str], nonce: str) -> UserInfo:
+        """Return an instance of UserInfo from token's ``id_token``."""
+        metadata = await self.load_metadata()
+        claims_params = {
+            "nonce": nonce,
+            "client_id": self._client_auth.client_id,
+        }
+        if "access_token" in token:
+            claims_params["access_token"] = token["access_token"]
+            claims_cls = CodeIDToken
+        else:
+            claims_cls = ImplicitIDToken
+
+        alg_values = metadata.get("id_token_signing_alg_values_supported", ["RS256"])
+
+        jwt = JsonWebToken(alg_values)
+
+        claim_options = {"iss": {"values": [metadata["issuer"]]}}
+
+        # Try to decode the keys in cache first, then retry by forcing the keys
+        # to be reloaded
+        jwk_set = await self.load_jwks()
+        try:
+            claims = jwt.decode(
+                token["id_token"],
+                key=jwk_set,
+                claims_cls=claims_cls,
+                claims_options=claim_options,
+                claims_params=claims_params,
+            )
+        except ValueError:
+            jwk_set = await self.load_jwks(force=True)  # try reloading the jwks
+            claims = jwt.decode(
+                token["id_token"],
+                key=jwk_set,
+                claims_cls=claims_cls,
+                claims_options=claim_options,
+                claims_params=claims_params,
+            )
+
+        claims.validate(leeway=120)  # allows 2 min of clock skew
+        return UserInfo(claims)
+
+    async def handle_redirect_request(self, request, client_redirect_url: bytes) -> str:
+        """Handle an incoming request to /login/sso/redirect
```

Comment:
Please expand upon what this is doing (in particular it is probably nice to mention that this stores the session in a cookie? I think that wasn't fully standard based on some conversation I saw fly past.)

Reply:
> I think OIDC is fine, putting it into a search engine / Wikipedia brings up pretty quickly how it is. It seems that OIDC is more generic than OAuth2? It might be nice to include (in the newsfragement?) that this includes OAuth2 providers though.

OAuth2 is a bunch of RFCs (mainly [6749](https://tools.ietf.org/html/rfc6749)) that define the authorization flows, but not where the endpoint should be, nor how user infos should work. OIDC defines that. For example, take a look at Google's OIDC discovery document, you'll see all the OAuth2 endpoints they have: https://accounts.google.com/.well-known/openid-configuration

> I'm not sure on this, I'm guessing that unlike SAML / CAS the data can't be pushed through with the client redirect?

Well it can, that's the purpose of the `state` parameter, to persist data throughout the flow. The signed cookie is here to be sure that the request is not intercepted by another malicious app/browser and prevent CSRF attacks. See [Auth0 docs about that](https://auth0.com/docs/protocols/oauth2/oauth-state), they say *“The cookie in which the state value is stored should be signed to prevent attackers to forge it.”*.

> This should work fine, we do this frequently. What was not working about it?

I forget to decorate the handlers with `@wrap_html_request_handler`, that's why it was not working. :)

> I suspect being consistent with the SAML implementation is beneficial. The SAML implementation (of a plug-in) probably provides more flexibility too.

> > * sync somehow the display name after the initial registration ; maybe periodically, maybe after each login
> > * get the user picture from the provider
> 
> I don't think we do this with SAML / CAS currently?

Nope, but that's definitely a thing we'd like to do in my company.

> One piece I see missing in here is handling the UI authentication process that was added in #5667. I think we'll definitely want this, but it might be OK to be added as a follow-up.

I'll look into it.

> Also, do you have any good resources for the flow diagram of OpenID Connect? I think it'd help me understand the various stages as I'm reviewing this! Thanks!

For the flow we are using (the authorization code flow), I think [this](https://auth0.com/docs/flows/concepts/auth-code) is a good resource to understand it. For some of the OIDC specific parts, maybe look at [this page](https://auth0.com/docs/protocols/oidc).

Thanks for this first review, I think I have a lot of comments/docstrings to add in various functions :D 
I'll work on it this afternoon/week-end.